File: 0652-find-duplicate-subtrees/0652-find-duplicate-subtrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    
public:
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        unordered_map<string, int> cnt;
        vector<TreeNode*> res;
        function<string(TreeNode*)> traverse = [&cnt, &res, &traverse](TreeNode* node) -> string{
            if (node == nullptr) return "";
            string representation = "(" + traverse(node -> left) + ")" + to_string(node -> val) + "(" + \
                                    traverse(node -> right) + ")";
            ++cnt[representation];
            if (cnt[representation] == 2) res.push_back(node);
            return representation;
        };
        
        traverse(root);
        return res;
    }
};

File: 0652-find-duplicate-subtrees/0652-find-duplicate-subtrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    triplet_to_id = dict()
    cnt = defaultdict(int)
    res = []
    def traverse(self, node: Optional[TreeNode]) -> int:
        if not node:
            return 0
        triplet = (self.traverse(node.left), node.val, self.traverse(node.right))
        if triplet not in self.triplet_to_id:
            self.triplet_to_id[triplet] = len(self.triplet_to_id) + 1
        id = self.triplet_to_id[triplet]
        self.cnt[id] += 1
        if self.cnt[id] == 2:
            self.res.append(node)
        return id
    
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        self.triplet_to_id = dict()
        self.cnt = defaultdict(int)
        self.res = []
        self.traverse(root)
        return self.res